src/BizHawk.Tests/Client.Common/lua/LuaTests.cs

```diff
@@ -0,0 +1,427 @@
+﻿using System;
+using System.Drawing;
+using System.Text;
+
+using BizHawk.Client.Common;
+using BizHawk.Common.StringExtensions;
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace BizHawk.Tests.Client.Common.Lua
+{
+	[TestClass]
+	public class LuaTests
+	{
+		private static readonly NLua.Lua LuaInstance = new();
+		private static readonly NLuaTableHelper _th = new(LuaInstance, Console.WriteLine);
+
+		// NOTE: NET6 changed Default encoding behavior.
+		// It no longer represents the default ANSI encoding
+		// Rather it now represents the "default encoding for this NET implementation"
+		// on Win10 this seems to just be an alias to UTF8, so it doesn't work here
+
+		public static string? FixString(string? s)
+			=> s is null
+				? null
+				: Encoding.UTF8.GetString(s.ToCharCodepointArray());
+
+		public static string? UnFixString(string? s)
+			=> s is null
+				? null
+				: StringExtensions.CharCodepointsToString(Encoding.UTF8.GetBytes(s));
+
+		private static object? ExpectedValue { get; set; }
+
+		[LuaMethod("pass_object", "")]
+		public static void PassObject(object? o)
+			=> Assert.IsTrue(o == ExpectedValue);
+
+		// seems nil passed over here gets turned into false
+		[LuaMethod("pass_bool", "")]
+		public static void PassBool(bool? o)
+			=> Assert.IsTrue(o == ((bool?)ExpectedValue ?? false));
+
+		[LuaMethod("pass_s8", "")]
+		public static void PassS8(sbyte? o)
+			=> Assert.IsTrue(o == (sbyte?)ExpectedValue);
+
+		[LuaMethod("pass_u8", "")]
+		public static void PassU8(byte? o)
+			=> Assert.IsTrue(o == (byte?)ExpectedValue);
+
+		[LuaMethod("pass_s16", "")]
+		public static void PassS16(short? o)
+			=> Assert.IsTrue(o == (short?)ExpectedValue);
+
+		[LuaMethod("pass_u16", "")]
+		public static void PassU16(ushort? o)
+			=> Assert.IsTrue(o == (ushort?)ExpectedValue);
+
+		[LuaMethod("pass_s32", "")]
+		public static void PassS32(int? o)
+			=> Assert.IsTrue(o == (int?)ExpectedValue);
+
+		[LuaMethod("pass_u32", "")]
+		public static void PassU32(uint? o)
+			=> Assert.IsTrue(o == (uint?)ExpectedValue);
+
+		[LuaMethod("pass_s64", "")]
+		public static void PassS64(long? o)
+			=> Assert.IsTrue(o == (long?)ExpectedValue);
+
+		[LuaMethod("pass_u64", "")]
+		public static void PassU64(ulong? o)
+			=> Assert.IsTrue(o == (ulong?)ExpectedValue);
+
+		[LuaMethod("pass_f32", "")]
+		public static void PassF32(float? o)
+			=> Assert.IsTrue(o == (float?)ExpectedValue);
+
+		[LuaMethod("pass_f64", "")]
+		public static void PassF64(double? o)
+			=> Assert.IsTrue(o == (double?)ExpectedValue);
+
+		[LuaMethod("pass_f128", "")]
+		public static void PassF128(decimal? o)
+			=> Assert.IsTrue(o == (decimal?)ExpectedValue);

```

Comment:
`decimal` is not [binary128](https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format)

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.decimal?view=netstandard-2.0
"The binary representation of a Decimal value is 128-bits consisting of a 96-bit integer number, and a 32-bit set of flags representing things such as the sign and scaling factor used to specify what portion of it is a decimal fraction."

Different significand/exponent width but it adds up to 128 bits in the end.